integrations/shopify/src/webhooks.ts

```diff
@@ -0,0 +1,179 @@
+import { IntegrationTaskKey, verifyRequestSignature } from "@trigger.dev/sdk";
+import { z } from "zod";
+import { Shopify, ShopifyRunTask } from "./index";
+import {
+  WebhookHeaderSchema,
+  WebhookSubscription,
+  WebhookSubscriptionDataSchema,
+  WebhookTopic,
+  WebhookTopicSchema,
+} from "./schemas";
+import { WebhookSource } from "@trigger.dev/sdk/triggers/webhook";
+
+export class Webhooks {
+  constructor(private runTask: ShopifyRunTask) {}
+
+  #apiUrl(client: NonNullable<Shopify["_client"]>) {
+    const { apiVersion, hostName } = client.config;
+    return new URL(`/admin/api/${apiVersion}/`, `https://${hostName}`);
+  }
+
+  // just here as an example if we ever want better platform support
+  #createWithFetch(
+    key: IntegrationTaskKey,
+    params: {
+      topic: WebhookTopic;
+      address: string;
+      fields?: string[];
+    }
+  ): Promise<WebhookSubscription> {
+    return this.runTask(
+      key,
+      async (client, task, io) => {
+        const resource = {
+          webhook: {
+            topic: params.topic,
+            address: params.address,
+            fields: params.fields,
+          },
+        };
+
+        const request = new Request(new URL("webhooks.json", this.#apiUrl(client)), {
+          method: "POST",
+          headers: {
+            "X-Shopify-Access-Token": client.config.adminApiAccessToken,
+            "Content-Type": "application/json",
+          },
+          body: JSON.stringify(resource),
+        });
+
+        const response = await fetch(request.clone());
+
+        if (!response.ok) {
+          await handleWebhookError("WEBHOOK_CREATE", request, response);
+        }
+
+        const webhook = await response.json();
+        const parsed = WebhookSubscriptionDataSchema.parse(webhook);
+
+        return parsed.webhook;
+      },
+      {
+        name: "Create Webhook with Fetch",
+        params,
+        properties: [
+          { label: "Webhook URL", text: params.address },
+          { label: "Topic", text: params.topic },
+        ],
+      }
+    );
+  }
+}
+
+export function createWebhookEventSource(integration: Shopify) {
+  return new WebhookSource({
+    id: "shopify",
+    schemas: {
+      params: z.object({
+        topic: WebhookTopicSchema,
+        fields: z.string().array().optional(),
+      }),
+      // config: z.record(z.string().array()),
+    },
+    version: "0.1.0",
+    integration,
+    key: (params) => `${params.topic}-${params.fields?.join(".")}`,
+    crud: {
+      create: async ({ io, ctx }) => {
+        const webhook = await io.integration.rest.Webhook.save("create-webhook", {
+          fromData: {
+            address: ctx.url,
+            topic: ctx.params.topic,
+            fields: ctx.params.fields,
+            // fields: ctx.config.desired.fields,
```

Comment:
`fields` is an optional array of property keys the payload should include. Includes all if not specified. Up for debate if this should be a param or config that may be merged across triggers. It shouldn't matter too much either way.